ibis/backends/datafusion/client.py

```diff
@@ -0,0 +1,204 @@
+import ibis.expr.operations as ops
+import ibis.expr.schema as sch
+import ibis.expr.types as ir
+from ibis.backends.base import Client, Database
+
+
+# TODO(kszucs): support nested and parametric types
+# consolidate with the logic from the parquet backend
+
+_to_ibis_dtypes = {
+    pa.int8(): dt.Int8,
+    pa.int16(): dt.Int16,
+    pa.int32(): dt.Int32,
+    pa.int64(): dt.Int64,
+    pa.uint8(): dt.UInt8,
+    pa.uint16(): dt.UInt16,
+    pa.uint32(): dt.UInt32,
+    pa.uint64(): dt.UInt64,
+    pa.float16(): dt.Float16,
+    pa.float32(): dt.Float32,
+    pa.float64(): dt.Float64,
+    pa.string(): dt.String,
+    pa.binary(): dt.Binary,
+    pa.bool_(): dt.Boolean,
+    pa.timestamp('ns'): dt.Timestamp,
+    pa.timestamp('ms'): dt.Timestamp
+}
+
+_to_pyarrow_types = {
+    dt.Int8: pa.int8(),
+    dt.Int16: pa.int16(),
+    dt.Int32: pa.int32(),
+    dt.Int64: pa.int64(),
+    dt.UInt8: pa.uint8(),
+    dt.UInt16: pa.uint16(),
+    dt.UInt32: pa.uint32(),
+    dt.UInt64: pa.uint64(),
+    dt.Float16: pa.float16(),
+    dt.Float32: pa.float32(),
+    dt.Float64: pa.float64(),
+    dt.String: pa.string(),
+    dt.Binary: pa.binary(),
+    dt.Boolean: pa.bool_(),
+    dt.Timestamp: pa.timestamp('ns'),
+}
+
+# TODO(kszucs): the following conversions are really rudimentary
+# we should have a pyarrow backend which would be responsible
+# for conversions between ibis types to pyarrow types
+
+def to_pyarrow_type(dtype):
+    #str(dtype))
+    try:
+        return _to_pyarrow_types[dtype.__class__]
+    except KeyError as e:
+        if isinstance(dtype, dt.Array):
+            return pa.list_(to_pyarrow_type(dtype.value_type))
+        raise e
+
+
+@dt.dtype.register(pa.DataType)
+def from_pyarrow_type(arrow_type, nullable=True):
+    return _to_ibis_dtypes[arrow_type](nullable=nullable)
+
+
+@sch.infer.register(pa.Schema)
+def infer_pyarrow_schema(schema):
+    fields = [(f.name, dt.dtype(f.type, nullable=f.nullable)) for f in schema]
+    return sch.schema(fields)
+
+
+class DataFusionTable(ops.DatabaseTable):
+    pass
+
+
+class DataFusionDatabase(ops.DatabaseTable):
+    pass
+
+
+class DataFusionClient(Client):
+    def __init__(self, backend, config):
+        self.backend = backend
+
+        if isinstance(config, df.ExecutionContext):
+            self.context = config
+        else:
+            self.context = df.ExecutionContext()
+            for name, path in config.items():
+                if path.endswith('.csv'):
+                    self.register_csv(name, path)
+                elif path.endswith('.parquet'):
+                    self.register_parquet(name, path)
+                else:
+                    raise ValueError('Wrong format')
+
+    def register_csv(self, name, path, schema=None):
+        self.context.register_csv(name, path, schema=schema)
+
+    def register_parquet(self, name, path, schema=None):
+        self.context.register_parquet(name, path, schema=schema)
+
+    # def database(self, name='public'):
```

Comment:
Maybe just delete this code?

Reply:
For other backends, this already warns with a `FutureWarning`, and the plan is to deprecate it, and make things simpler.

Reply:
What is the rationale for deprecating this method?

Reply:
I think the `Database` class doesn't add much value. Its methods are duplicated in `Client`, and just makes things more complex.

For example, I think we can just have:
```python
con = ibis.sqlite.connect(...)
con.drop_database('foo')
```

and this doesn't seem necessary to me:
```python
con = ibis.sqlite.connect(...)
db = con.database('foo')
db.drop()
```

Does it sound reasonable?

Reply:
Sure, in the drop case that makes sense, but there are other use cases like cross-database joins where I think an object is warranted.

Generally agree that there's too much duplication of APIs though, so it's fine IMO to pare things down.

In that case we shouldn't be afraid to introduce breaking changes and leverage semantic versioning instead of deprecating (and removing the API at some arbitrarily later date).

Reply:
I think cross-database joins should be possible with `con.table('foo', database='bar')`, right? If I'm missing something, happy to reconsider, but for the use cases I'm aware of, the whole `Database` class is just duplication of `Client`.

If there is some special case where is needed, I think making `.database()` return an instance of `Client` would make more sense.

Regarding deprecation, my plan is to show the `FutureWarning` in Ibis 2.* and have the feature removed in Ibis 3.*. Happy to discuss and hear other options if you have better ideas.

Reply:
Why not remove it when bumping to version 2? That's a major version bump, and that's where a breaking API change is allowed and should occur.

Reply:
The idea (and my understanding on how semver is supposed to be implemented) is next:

If an Ibis 1.* user has this code:
```python
import ibis

con = ibis.sqlite.connect(...)
db = con.database()
```

This is fine in Ibis 1.* (works and no warning is issued). If we do what you say, and the user uprades Ibis and runs this code in Ibis 2.*, the code won't work.

If instead, we add a `FutureWarning`, when the user upgrades to Ibis 2.0, the code will still work, but will start raising the warning. When the user wants to upgrade to 3.0, they should know that the code will still work, unless there is any `FutureWarning` in the logs.

Do you think we should do this for minor instead of major releases?